text/0045-add-notion-of-protocol.md

```diff
@@ -0,0 +1,144 @@
+
+Adding these functions to the Services API:
+- `const char *obs_service_get_connect_info(uint32_t type, const obs_service_t *service)`: return the connection information related to the given type (list above).
+- `bool obs_service_can_try_to_connect(const obs_service_t *service)`: return if the service can try to connect.
+  - return `bool (*can_try_to_connect)(void *data)` result.
+  - return true if `bool (*can_try_to_connect)(void *data)` is not implemented in the service.
+  - return false if the service object is `NULL`.
+
+`obs_service_get_url()`, `obs_service_get_key()`, `obs_service_get_username()`, and `obs_service_get_password()` will be deprecated in favor of `obs_service_get_connect_info()`.
+
+In the future, if we want to support a protocol that doesn't use a server URL, this scenario is covered by this design.
+
+### About `rtmp-services`
+
+`rtmp-services` is a plugin that provides two services `rtmp_common` and `rtmp_custom`. The use of "rtmp" in the naming no longer means that it only supports RTMP, it was and is kept to avoid breakage.
+
+- In `rtmp_common`, services must at least support H264 as video codec and AAC or Opus as audio codec. This is a limitation required by the simple output mode.
+- The service `rtmp_custom` will depend on protocol prefixes from outputs types to detect which protocol is in use.
+
+#### About service JSON file
+
+- For `services` objects, a `"protocol"` field will be added. This will be used to indicate the protocol for services in situation where:
+  - The protocol can not be deduced from servers URL prefix.
+  - The services is RTMP but has a HTTP url as its server URL have custom code in the plugin will have the protocol field enforced to RTMP.
+- Services that use a protocol that is not registered will not be shown. e.g. OBS Studio without RTMPS support will not show services and servers that rely on RTMPS.
+- Codecs field for audio and video will be added to allow services to limit which codec is compatible with the service.
+- `"output"` field in the `"recommended"` object will be deprecated, but it will be kept for backward compatibility. `const char *(*get_output_type)(void *data)` in `obs_service_info` will be no longer used by `rtmp-services`.
+  - The JSON schema will be modified to require the `"protocol"`field when the protocol is not auto-detectable. The same for the `"output"` field to keep backward compatibility.
+
+## UI
+
+### Custom server protocol auto-detection
+
+The user can set a custom server through the service "Custom…", the protocol will be deduced based on URL prefixes from service outputs and default to RTMP if no matches.
+
+NOTE: Allowing the user to manually choose the protocol (e.g. HLS) is part of RFC 39.
+
+### Audio encoders
+
+If the service/output is compatible with more than one audio codec, the user should be able to choose in the UI which one:
```

Comment:
As a user could I set my preference order? `AV1 -> H265 -> H264`. After the session is negotiated OBS takes the best one possible.

Will this also be influenced by the remote and its feature capability
* If remote supports SVC use `X`
* If remote supports Simulcast use `Y`

Reply:
The user only set one encoder (e.g. NVENC H264, AOM AV1… ) in the settings, only codecs supported by the service and the outputs are shown in the settings UI.

The user (auto-wizard) decides which encoder (one codec) it uses, the UI and the output **must** follow user settings.

This is not part of the RFC, this is how OBS Studio actually works with a hard-coded codecs list.
The RFC just plan to replace the list with service supported codecs.